### PR TITLE
Don't create intermediate list to add indexable fields

### PR DESCRIPTION
**Pull request by mfussenegger**
No measurable improvement using end-to-end benchmarks, they fluctuate +- ~5% even with no change, but allocation rate is consistently down:

```
Q: insert into id_int_value_str ("id", "value") values ($1, $2)
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |        9.521 ±   11.875 |      6.131 |      8.100 |      9.318 |    349.078 |
|   V2    |        9.295 ±   11.156 |      6.184 |      7.779 |      9.044 |    321.629 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               -   2.41%                           -   4.04%
There is a 33.99% probability that the observed difference is not random, and the best estimate of that difference is 2.41%
The test has no statistical significance

Q: insert into id_int_value_str ("id", "value") values ($1, $2)
C: 10
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |       13.364 ±   11.161 |      7.782 |     11.268 |     12.169 |    126.731 |
|   V2    |       14.005 ±    7.832 |      8.004 |     12.009 |     13.485 |     79.683 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               +   4.69%                           +   6.36%
There is a 86.30% probability that the observed difference is not random, and the best estimate of that difference is 4.69%
The test has no statistical significance

Q: insert into tbl ("id", "value") values ($1, $2)
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |        7.918 ±    2.399 |      6.126 |      7.490 |      8.191 |     50.352 |
|   V2    |        7.627 ±    2.714 |      5.154 |      7.128 |      7.843 |     40.551 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               -   3.75%                           -   4.96%
There is a 98.91% probability that the observed difference is not random, and the best estimate of that difference is 3.75%
The test has statistical significance

Q: insert into tbl ("id", "value") values ($1, $2)
C: 10
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |       20.825 ±   15.453 |      7.146 |     16.438 |     20.896 |    139.556 |
|   V2    |       18.971 ±    9.829 |      7.148 |     16.023 |     19.086 |    108.264 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               -   9.32%                           -   2.56%
There is a 99.86% probability that the observed difference is not random, and the best estimate of that difference is 9.32%
The test has statistical significance


System/JVM Metrics (durations in ms, byte-values in MB)
    |    YOUNG GC            |       OLD GC           |      HEAP         |     ALLOC
    |  cnt      avg      max |  cnt      avg      max |  initial     used |     rate      total
 V1 |  143     3.93     5.97 |    0     0.00     0.00 |      268     1429 |   783.08      46813
 V2 |  136     4.04     4.79 |    0     0.00     0.00 |      268      948 |   758.50      45222
```

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/GeoShapeFieldMapper.java [hunk 1]
```diff
@@ -404,16 +404,17 @@ public void parse(ParseContext context) throws IOException {
...
         Field[] indexableFields = fieldType().defaultStrategy().createIndexableFields(shape);
-        List<IndexableField> fields = new ArrayList<>(Arrays.asList(indexableFields));
-        createFieldNamesField(context, fields);
-        for (IndexableField field : fields) {
-            context.doc().add(field);
+        for (var field : indexableFields) {
+            doc.add(field);
         }
+        Consumer<IndexableField> addField = doc::add;
+        createFieldNamesField(context, addField);
```
Comment (matriv) [inline server/src/main/java/org/elasticsearch/index/mapper/GeoShapeFieldMapper.java:413]: Is there a difference if you inline it?

**Review by matriv — APPROVED** [review]
nice! left a question about inlining the consumer

File: server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java [hunk 2]
```diff
@@ -151,12 +153,10 @@ public CopyTo copyTo() {
...
     public void parse(ParseContext context) throws IOException {
-        final List<IndexableField> fields = new ArrayList<>(2);
+        Document doc = context.doc();
+        Consumer<IndexableField> addField = field -> doc.add(field);
         try {
-            parseCreateField(context, fields);
-            for (IndexableField field : fields) {
-                context.doc().add(field);
-            }
+            parseCreateField(context, addField);
```
Comment (matriv) [inline server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java:159]: same here

File: server/src/main/java/org/elasticsearch/index/mapper/GeoShapeFieldMapper.java (hunk 1 above)
Comment (mfussenegger) [reply server/src/main/java/org/elasticsearch/index/mapper/GeoShapeFieldMapper.java:413]: No, I don't think it does.

**Pull request merged**